### PR TITLE
Creating bag of words classifier

### PR DESCRIPTION
**Pull request by Lam7150**
Resolves issue #24 #26 

* bag of words classifier class

### CODE REVIEW

File: classification/BOWClassifier.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
+import json
+import numpy as np
+import sys
+
+sys.path.append("..")
+from utils.dataset_db import dynamo_db
```
Comment (arynh) [inline classification/BOWClassifier.py]: poof then dynamo_db was gone

File: classification/BOWClassifier.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+
+def main():
```
Comment (arynh) [inline classification/BOWClassifier.py]: samples can be written into jupyter notebook for  a e s t h e t i c s  ✨ and good examples

File: classification/BOWClassifier.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+    bow = BOWClassifier("trained.json", "categories_dict.json")
+    bow.predict_set(dynamo_db.get_dataset("organizations_text"))
+    bow.save_predictions("predictions.json")
+    bow.load_predictions("predictions.json")
+    bow.load_targets(json.load(open("trained.json")))
+    print(bow.get_f1_score())
+
+
+class BOWClassifier:
```
Comment (JoshBurke) [inline classification/BOWClassifier.py:6]: Add class-level docstring

File: classification/BOWClassifier.py [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+    themes = {}
```
Comment (JoshBurke) [inline classification/BOWClassifier.py]: Describe member variables in class-level docstring or inline comments

File: classification/BOWClassifier.py (hunk 3 above)
Comment (JoshBurke) [reply classification/BOWClassifier.py:6]: https://www.python.org/dev/peps/pep-0257/

File: classification/BOWClassifier.py [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+    training_data = {}
+    dictionary = None
+
+    predictions = None
+
+    testing_data = None
+    testing_targets = None
+
+    """
+    Initializes bag of words classifier object
```
Comment (JoshBurke) [inline classification/BOWClassifier.py]: Function docstrings go directly after function signature

File: classification/BOWClassifier.py [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+    
+    :param str train_data: training data filename
+    :param str dict_data: dictionary of category words filename
+    """
+
+    def __init__(self, train_data: str, dict_data: str):
```
Comment (JoshBurke) [inline classification/BOWClassifier.py]: Specify the format of these inputs in readme

File: classification/BOWClassifier.py [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+        with open(train_data, "r") as input_file:
```
Comment (JoshBurke) [inline classification/BOWClassifier.py]: do this from memory instead of from file

File: classification/BOWClassifier.py [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+            )
+
+            category_f1_scores[theme_name] = f1
+
+        return np.mean(np.array(org_f1_scores)), category_f1_scores
+
+    """
+    Load the testing targets for testing dataset
+
+    :param dict target_data: dictionary containing target data for testing dataset
```
Comment (arynh) [inline classification/BOWClassifier.py]: describe what a target is
e.g. how does the mapping work?

File: classification/BOWClassifier.py [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+                    if j in testing_targets[i]:
+                        tp += 1
+                    else:
+                        fp += 1
+                else:
+                    if j in testing_targets[i]:
+                        fn += 1
+                    else:
+                        tn += 1
+            precision = 0 if tp + fp == 0 else tp / (tp + fp)
```
Comment (arynh) [inline classification/BOWClassifier.py]: parens might be enccessaary here

**Review by arynh — DISMISSED** [review]

File: classification/README.md [hunk 10]
```diff
@@ -10,5 +10,17 @@ Each has its own strengths and weaknesses, blah blah blah
 ## Stochastic Gradient Descent
 Something something
 
-## Bag of Words
-something something
+# Bag of Words Classifier
```
Comment (JoshBurke) [inline classification/README.md]: This should be in h2 (two #'s)

File: classification/README.md [hunk 11]
```diff
@@ -10,5 +10,17 @@ Each has its own strengths and weaknesses, blah blah blah
...
-## Bag of Words
-something something
+# Bag of Words Classifier
+Bag of words is a statistical method of classifying documents based off the frequency and relevance of words in a document's text. In this implementation, given a organization's website text, using bag of words algorithm, the classifer will predict the themes of the organization.
+
+Bag of words works in two steps:
+1. Generate a dictionary of words with tf-idf scores for each category for the classifier to reference
+2. Use the bag of words algorithm to classify organizations by counting/scoring the words contained in the website's text. The highest scored categories are the predicted classifications given to the organization
+
+# What is TF-IDF?
```
Comment (JoshBurke) [inline classification/README.md]: this should also be h2

File: classification/README.md [hunk 12]
```diff
@@ -10,5 +10,17 @@ Each has its own strengths and weaknesses, blah blah blah
...
+Tf-idf is a weighting scheme that assigns each term in a document a weight based on its term frequency (tf) and inverse document frequency (idf). Tf-idf is especially valuable in the bag of words algorithm to increase the accuracy of its predictions by scoring words in an organization's text with a weighted score accounting for its relevance to each category, rather than simply summing the number of words that appear in each category.
+
+For example, the word "peacock" is more rare compared to the word "girl" across all categories. This rarity is accounted for by the "idf" in tf-idf scoring to make clear that it is a stronger indicator of a specific category, which in this case would be animals. The word "peacock" would also be more common in websites with the theme "animal", rather than "economic development", accounted for by the "tf" in tf-idf scoring to indicate that multiple occurences of the same word in a given category will incline a prediction of that category.
+
+# Important things to note
```
Comment (JoshBurke) [inline classification/README.md]: this should be in h2 too

File: classification/BOWClassifier.py (hunk 9 above)
Comment (JoshBurke) [reply classification/BOWClassifier.py]: if (tp + fp == 0)

**Comment by JoshBurke** [comment]
Great stuff

**Pull request merged**